Tools/HLS_Arbitrary_Precision_Types/include/ap_impl/ap_private.h

```diff
@@ -0,0 +1,7200 @@
+
+template <int _AP_N8>
+struct valtype<_AP_N8, false> {
+  typedef uint64_t Type;
+};
+
+template <int _AP_N8>
+struct valtype<_AP_N8, true> {
+  typedef int64_t Type;
+};
+
+template <>
+struct valtype<1, false> {
+  typedef unsigned char Type;
+};
+template <>
+struct valtype<2, false> {
+  typedef unsigned short Type;
+};
+template <>
+struct valtype<3, false> {
+  typedef unsigned int Type;
+};
+template <>
+struct valtype<4, false> {
+  typedef unsigned int Type;
+};
+template <>
+struct valtype<1, true> {
+  typedef signed char Type;
+};
+template <>
+struct valtype<2, true> {
+  typedef short Type;
+};
+template <>
+struct valtype<3, true> {
+  typedef int Type;
+};
+template <>
+struct valtype<4, true> {
+  typedef int Type;
+};
+
+template <bool enable>
+struct ap_private_enable_if {};
+template <>
+struct ap_private_enable_if<true> {
+  static const bool isValid = true;
+};
+
+// When bitwidth < 64
+template <int _AP_W, bool _AP_S>
+class ap_private<_AP_W, _AP_S, true> {
+  // SFINAE pattern.  Only consider this class when _AP_W <= 64
+  const static bool valid = ap_private_enable_if<_AP_W <= 64>::isValid;
+
+#ifdef _MSC_VER
+#pragma warning(disable : 4521 4522)
+#endif
+ public:
+  typedef typename valtype<(_AP_W + 7) / 8, _AP_S>::Type ValType;
+  typedef ap_private<_AP_W, _AP_S> Type;
+  template <int _AP_W2, bool _AP_S2>
+  struct RType {
+    enum {
+      mult_w = _AP_W + _AP_W2,
+      mult_s = _AP_S || _AP_S2,
+      plus_w =
+          AP_MAX(_AP_W + (_AP_S2 && !_AP_S), _AP_W2 + (_AP_S && !_AP_S2)) + 1,
+      plus_s = _AP_S || _AP_S2,
+      minus_w =
+          AP_MAX(_AP_W + (_AP_S2 && !_AP_S), _AP_W2 + (_AP_S && !_AP_S2)) + 1,
+      minus_s = true,
+      div_w = _AP_W + _AP_S2,
+      div_s = _AP_S || _AP_S2,
+      mod_w = AP_MIN(_AP_W, _AP_W2 + (!_AP_S2 && _AP_S)),
+      mod_s = _AP_S,
+      logic_w = AP_MAX(_AP_W + (_AP_S2 && !_AP_S), _AP_W2 + (_AP_S && !_AP_S2)),
+      logic_s = _AP_S || _AP_S2
+    };
+    typedef ap_private<mult_w, mult_s> mult;
+    typedef ap_private<plus_w, plus_s> plus;
+    typedef ap_private<minus_w, minus_s> minus;
+    typedef ap_private<logic_w, logic_s> logic;
+    typedef ap_private<div_w, div_s> div;
+    typedef ap_private<mod_w, mod_s> mod;
+    typedef ap_private<_AP_W, _AP_S> arg1;
+    typedef bool reduce;
+  };
+  enum { APINT_BITS_PER_WORD = sizeof(uint64_t) * 8 };
+  enum {
+    excess_bits = (_AP_W % APINT_BITS_PER_WORD)
+                      ? APINT_BITS_PER_WORD - (_AP_W % APINT_BITS_PER_WORD)
+                      : 0
+  };
+  static const uint64_t mask = ((uint64_t)~0ULL >> (excess_bits));
+  static const uint64_t not_mask = ~mask;
+  static const uint64_t sign_bit_mask = 1ULL << (APINT_BITS_PER_WORD - 1);
+  template <int _AP_W1>
+  struct sign_ext_mask {
+    static const uint64_t mask = ~0ULL << _AP_W1;
+  };
+  static const int width = _AP_W;
+
+  enum {
+    BitWidth = _AP_W,
+    _AP_N = 1,
+  };
+  ValType VAL; ///< Used to store the <= 64 bits integer value.
```

Comment:
```suggestion
  ValType VAL{0}; ///< Used to store the <= 64 bits integer value.
```

Reply:
Oh wait nvm I see you added it in the way it was the PR from Shahzad, I guess that should work out too